### PR TITLE
Allow building without the daemon

### PR DESCRIPTION
**Pull request by lantw44**
The daemon part of udisks is specific to Linux, but the client library
and tools built on it are likely to be portable to other operating
systems because they consist of mostly wrappers of the D-Bus API. While
the official udisks daemon does not run on non-Linux systems, it is
possible for other people to create an unofficial service which is
mostly compatible with udisks for non-Linux systems.

This commit is intended to support gnome-control-center on FreeBSD.
gnome-control-center uses udisks client library to obtain information
for showing disk size in its 'About' panel. On FreeBSD, there is an
unofficial and incomplete implementation of udisks service called
'bsdisks', which is already good enough for this use case.

This pull request also include commits to avoid hardcoding `/bin/bash` in Makefile and remove usage of glibc-specific `__GNUC_PREREQ` macro. Both are required to get the client library built on FreeBSD.

References:
https://gitlab.gnome.org/GNOME/gnome-control-center/issues/606
https://gitlab.gnome.org/GNOME/gnome-control-center/merge_requests/513
https://bitbucket.org/arrowd/bsdisks/issues/2
https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=240138

### CODE REVIEW

**Comment by pauamma** [comment]
I'm getting the following error:
`doc/Makefile.am:76: error: ENABLE_GTK_DOC does not appear in AM_CONDITIONAL
automake: error: cannot open < gtk-doc.make: No such file or directory
autoreconf-2.69: automake failed with exit status: 1
*** Error code 1`
perhaps due to
`--- a/doc/Makefile.am
+++ b/doc/Makefile.am
@@ -73,7 +73,9 @@ extra_files =                                                                  \
 
 # ------------------------------------------------------------------------------
 
+if ENABLE_GTK_DOC
 all-local: generate-main-sgml-file
+endif`

Also, to be consistent with the intent of the above, I believe you also need something like
`run_gtkdocize="yes"
for arg
do
        if [ "$arg" = "--disable-gtk-doc" ]; then
                run_gtkdocize="no"
                break
        fi
done
if [ "$run_gtkdocize" = "yes" ]; then
        gtkdocize --copy || exit 1
fi`
in autogen.sh, otherwise you still need gtk-doc installed.

**Comment by vojtechtrefny** [comment]
Jenkins, ok to test.

**Comment by lantw44** [comment]
> I'm getting the following error:
> 
> > `doc/Makefile.am:76: error: ENABLE_GTK_DOC does not appear in AM_CONDITIONAL`
> > `automake: error: cannot open < gtk-doc.make: No such file or directory`
> > `autoreconf-2.69: automake failed with exit status: 1`
> > `*** Error code 1`

You have to install gtk-doc. `--disable-gtk-doc` tells the configure script how the build should be configured. It doesn't affect how the configure script itself is generated.

> perhaps due to
> 
> > `--- a/doc/Makefile.am`
> > `+++ b/doc/Makefile.am`
> > `@@ -73,7 +73,9 @@ extra_files =                                                                  \`
> > ` `
> > ` # ------------------------------------------------------------------------------`
> > ` `
> > `+if ENABLE_GTK_DOC`
> > ` all-local: generate-main-sgml-file`
> > `+endif`

The reason i added `if ENABLE_GTK_DOC` was that `generate-main-sgml-file` uses the non-portable `sed -i` syntax which isn't compatible with FreeBSD `sed`. Rather than spending more time fixing gtk-doc issues, I thought it is better to have a minimal patch to see if the upstream is going to accept it.

> 
> Also, to be consistent with the intent of the above, I believe you also need something like
> 
> > `run_gtkdocize="yes"`
> > `for arg`
> > `do`
> > `        if [ "$arg" = "--disable-gtk-doc" ]; then`
> > `                run_gtkdocize="no"`
> > `                break`
> > `        fi`
> > `done`
> > `if [ "$run_gtkdocize" = "yes" ]; then`
> > `        gtkdocize --copy || exit 1`
> > `fi`
> > (only with proper indentation, which GitHub Markdown apparently doesn't want me to have) in autogen.sh, otherwise you still need gtk-doc installed.

No, gtk-doc is only optional when you have configure script and Makefile.in pre-generated, such as building from a release tarball. If you have to run `autogen.sh` because you build from git, you have to install gtk-doc.

**Comment by vojtechtrefny** [comment]
Looks good to me in general, but I didn't test this beyond running `./autogen.sh --disable-daemon --disable-gtk-doc && make` which looks ok, the library was build, but I'd like to see someone testing this with the BSD daemon first.

@pauamma You started the mailing list thread, can you confirm this is working for you?

**Comment by lantw44** [comment]
>  I'd like to see someone testing this with the BSD daemon first.

I only tested `udisksctl dump` and `udisksctl info` when I wrote the patch. I also found a few problems in bsdisks implementation, and the author of bsdisks has fixed it in version 0.9.

This is what `udisksctl dump` shows on FreeBSD.
```
/org/freedesktop/UDisks2/block_devices/ada0:
  org.freedesktop.UDisks2.Block:
    Configuration:              []
    CryptoBackingDevice:        '/'
    Device:                     
    DeviceNumber:               0
    Drive:                      '/org/freedesktop/UDisks2/drives/ada0'
    HintAuto:                   false
    HintIconName:               
    HintIgnore:                 false
    HintName:                   
    HintPartitionable:          false
    HintSymbolicIconName:       
    HintSystem:                 false
    Id:                         Hitachi HDS721050CLA662_JP1532FN1WTEUK
    IdLabel:                    
    IdType:                     
    IdUUID:                     
    IdUsage:                    
    IdVersion:                  
    MDRaid:                     '/'
    MDRaidMember:               '/'
    PreferredDevice:            
    ReadOnly:                   false
    Size:                       500107862016
    Symlinks:                   
  org.freedesktop.UDisks2.PartitionTable:
    Type:               gpt

/org/freedesktop/UDisks2/block_devices/ada0p1:
  org.freedesktop.UDisks2.Block:
    Configuration:              []
    CryptoBackingDevice:        '/'
    Device:                     
    DeviceNumber:               0
    Drive:                      '/org/freedesktop/UDisks2/drives/ada0p1'
    HintAuto:                   false
    HintIconName:               
    HintIgnore:                 false
    HintName:                   
    HintPartitionable:          false
    HintSymbolicIconName:       
    HintSystem:                 false
    Id:                         Hitachi HDS721050CLA662_JP1532FN1WTEUK_p1
    IdLabel:                    
    IdType:                     
    IdUUID:                     
    IdUsage:                    
    IdVersion:                  
    MDRaid:                     '/'
    MDRaidMember:               '/'
    PreferredDevice:            
    ReadOnly:                   false
    Size:                       65536
    Symlinks:                   
  org.freedesktop.UDisks2.Partition:
    Flags:              0
    IsContained:        false
    IsContainer:        false
    Name:               
    Number:             1
    Offset:             20480
    Size:               65536
    Table:              '/org/freedesktop/UDisks2/block_devices/ada0'
    Type:               freebsd-boot
    UUID:

/org/freedesktop/UDisks2/block_devices/ada0p2:
  org.freedesktop.UDisks2.Block:
    Configuration:              []
    CryptoBackingDevice:        '/'
    Device:                     
    DeviceNumber:               0
    Drive:                      '/org/freedesktop/UDisks2/drives/ada0p2'
    HintAuto:                   false
    HintIconName:               
    HintIgnore:                 false
    HintName:                   
    HintPartitionable:          false
    HintSymbolicIconName:       
    HintSystem:                 false
    Id:                         Hitachi HDS721050CLA662_JP1532FN1WTEUK_p2
    IdLabel:                    
    IdType:                     
    IdUUID:                     
    IdUsage:                    
    IdVersion:                  
    MDRaid:                     '/'
    MDRaidMember:               '/'
    PreferredDevice:            
    ReadOnly:                   false
    Size:                       4294967296
    Symlinks:                   
  org.freedesktop.UDisks2.Filesystem:
    MountPoints:        
  org.freedesktop.UDisks2.Partition:
    Flags:              0
    IsContained:        false
    IsContainer:        false
    Name:               
    Number:             2
    Offset:             86016
    Size:               4294967296
    Table:              '/org/freedesktop/UDisks2/block_devices/ada0'
    Type:               freebsd-ufs
    UUID:  

/org/freedesktop/UDisks2/block_devices/ada0p3:
  org.freedesktop.UDisks2.Block:
    Configuration:              []
    CryptoBackingDevice:        '/'
    Device:                     
    DeviceNumber:               0
    Drive:                      '/org/freedesktop/UDisks2/drives/ada0p3'
    HintAuto:                   false
    HintIconName:               
    HintIgnore:                 false
    HintName:                   
    HintPartitionable:          false
    HintSymbolicIconName:       
    HintSystem:                 false
    Id:                         Hitachi HDS721050CLA662_JP1532FN1WTEUK_p3
    IdLabel:                    
    IdType:                     
    IdUUID:                     
    IdUsage:                    
    IdVersion:                  
    MDRaid:                     '/'
    MDRaidMember:               '/'
    PreferredDevice:            
    ReadOnly:                   false
    Size:                       42949672960
    Symlinks:                   
  org.freedesktop.UDisks2.Filesystem:
    MountPoints:        
  org.freedesktop.UDisks2.Partition:
    Flags:              0
    IsContained:        false
    IsContainer:        false
    Name:               
    Number:             3
    Offset:             0
    Size:               42949672960
    Table:              '/org/freedesktop/UDisks2/block_devices/ada0'
    Type:               freebsd-ufs
    UUID:  
...
/org/freedesktop/UDisks2/drives/ada0:
  org.freedesktop.UDisks2.Drive:
    CanPowerOff:                false
    Configuration:              {}
    ConnectionBus:              
    Ejectable:                  false
    Id:                         
    Media:                      
    MediaAvailable:             false
    MediaChangeDetected:        false
    MediaCompatibility:         
    MediaRemovable:             false
    Model:                      
    Optical:                    false
    OpticalBlank:               false
    OpticalNumAudioTracks:      0
    OpticalNumDataTracks:       0
    OpticalNumSessions:         0
    OpticalNumTracks:           0
    Removable:                  false
    Revision:                   
    RotationRate:               0
    Seat:                       
    Serial:                     JP1532FN1WTEUK
    SiblingId:                  
    Size:                       500107862016
    SortKey:                    
    TimeDetected:               0
    TimeMediaDetected:          0
    Vendor:                     Hitachi HDS721050CLA662
    WWN:      

/org/freedesktop/UDisks2/drives/ada1:
  org.freedesktop.UDisks2.Drive:
    CanPowerOff:                false
    Configuration:              {}
    ConnectionBus:              
    Ejectable:                  false
    Id:                         
    Media:                      
    MediaAvailable:             false
    MediaChangeDetected:        false
    MediaCompatibility:         
    MediaRemovable:             false
    Model:                      
    Optical:                    false
    OpticalBlank:               false
    OpticalNumAudioTracks:      0
    OpticalNumDataTracks:       0
    OpticalNumSessions:         0
    OpticalNumTracks:           0
    Removable:                  false
    Revision:                   
    RotationRate:               0
    Seat:                       
    Serial:                     GEC534RJ0ZN2PE
    SiblingId:                  
    Size:                       500107862016
    SortKey:                    
    TimeDetected:               0
    TimeMediaDetected:          0
    Vendor:                     Hitachi HDP725050GLA360
    WWN:  
```

**Comment by vojtechtrefny** [comment]
> Are there any other (udisksctl-based) tests we should perform?

Our test suite uses either the DBus API or Python bindings for the library, so there are not useful here. I think that if everything is working for you in FreeBSD, we can merge this.

**Comment by vpodzime** [comment]
Jenkins, test this, please.

**Comment by lantw44** [comment]
Is it ready to merge?

**Comment by tbzatek** [comment]
> Is it ready to merge?

Let me do some more tests...

**Comment by lantw44** [comment]
> Meaning if somebody wants to write a BSD-specific daemon part living in the udisks source tree, the technical background is ready for that.

If somebody is going to do it for FreeBSD, beware that FreeBSD libgeom, the library used for disk management, conflicts with GLib. FreeBSD libgeom and GLib include functions with the same names, so an executable or a library should not link to both libraries directly.

> Anyway, there are couple of build failures when compiled with `--disable-modules --enable-gtk-doc --disable-daemon` that should be fixed first:

I know `--disable-gtk-doc` is broken when `--disable-daemon` is used. Patches I posted here are just minimum required changes to build the client library without the daemon. Is it required to fix it in this pull request?

**Comment by tbzatek** [comment]
> If somebody is going to do it for FreeBSD, beware that FreeBSD libgeom, the library used for disk management, conflicts with GLib. FreeBSD libgeom and GLib include functions with the same names, so an executable or a library should not link to both libraries directly.

Ah, the `g_` namespace prefix clash. We do already pull too many libraries in the process through modules, conflicts are not entirely avoidable.

> I know `--disable-gtk-doc` is broken when `--disable-daemon` is used. Patches I posted here are just minimum required changes to build the client library without the daemon. Is it required to fix it in this pull request?

Yes please, we don't want to ship half-baked work. There's still plenty of time before the 2.9.0 release. The code looks good, it's just the docs. It might be worth separating the daemon symbols and sections the similar way it's done for modules.

**Comment by lantw44** [comment]
I pushed 3 more commits to fix gtk-doc build and out-of-source build.

File: doc/Makefile.am [hunk 1]
```diff
@@ -88,60 +88,60 @@ generate-main-sgml-file: $(DOC_MAIN_SGML_FILE).in $(UDISKS_SECTIONS).in $(UDISKS
 	cp $(UDISKS_SECTIONS).in $(builddir)/$(UDISKS_SECTIONS) && \
 	cp $(UDISKS_TYPES).in $(builddir)/$(UDISKS_TYPES)
 if HAVE_LVM2
-	$(AM_V_at) $(SED) -i -f $(DOC_DIR)/$(DOC_MAIN_SGML_FILE).lvm2.sed $(builddir)/$(DOC_MAIN_SGML_FILE)
-	$(AM_V_at) $(SED) -i -f $(DOC_DIR)/$(UDISKS_SECTIONS).lvm2.sed $(builddir)/$(UDISKS_SECTIONS)
-	$(AM_V_at) $(SED) -i -f $(DOC_DIR)/$(UDISKS_TYPES).lvm2.sed $(builddir)/$(UDISKS_TYPES)
+	$(AM_V_at) $(SED) -f $(DOC_DIR)/$(DOC_MAIN_SGML_FILE).lvm2.sed $(builddir)/$(DOC_MAIN_SGML_FILE) > $(builddir)/$(DOC_MAIN_SGML_FILE).tmp && mv $(builddir)/$(DOC_MAIN_SGML_FILE).tmp $(builddir)/$(DOC_MAIN_SGML_FILE)
```
[the remark this answers is not in the document]
Comment (lantw44) [reply doc/Makefile.am]: There are also a few `mv` in `udisks/Makefile.am`. Do you want to change them as well?

[the remark this answers is not in the document]
Comment (tbzatek) [reply doc/Makefile.am]: Good catch! Although this never caused an issue, it wouldn't harm to fortify that as well.

**Comment by lantw44** [comment]
I updated the pull request to address all issues mentioned above.

**Comment by arrowd** [comment]
> FWIW, the daemon classes are quite well structured and from the objective nature can be interchanged with different implementation. Meaning if somebody wants to write a BSD-specific daemon part living in the udisks source tree, the technical background is ready for that. 

I looked into that a bit and while UDisks itself is indeed structured well, it uses a lot of Linux-specific dependencies. I started with blockdev and there are two options to move forward:
- Make it an optional dependency for UDisks
- Try porting it to FreeBSD, which sends us down to 

```
Header file linux/random.h not found.
Package 'libkmod' not found
Package 'libcryptsetup' not found
libvolume_key.h not available
keyutils.h not available
Package 'devmapper' not found
Package 'fdisk' not found
Package 'mount' not found
Package 'libparted' not found
Package 'bytesize' not found
Package 'libndctl' not found
Package 'libnvme' not found
Package 'libatasmart' not found
```

which probably turn out to also be a very Linux-specific bits that are hard to get ported.

It is a bit of chicken-and-egg problem - to start porting UDisks to an another OS you first need to strip all Linux specific-functionality, but there is no available switch for that at the build system level. But adding such a switch only makes sense if there is at least one alternative OS that UDisks may run on, and currently there is none.

IIRC, this is the main reason I decided to write bsdisks from scratch, rather than trying to port UDisks to FreeBSD. And while it is clear that bsdisks is no match to UDisks in terms of functionality and stability, it still feels to me that porting UDisks is an unrealistic task.

**Comment by arrowd** [comment]
> That said, if you need to clarify some parts of the API or extend it for BSD-specific stuff, feel free to open separate tickets here. 

Sure thing, thank you. We do try to not deviate from the original UDisks as much as possible. This PR was created exactly because of this, which made it possible to reuse libudisks as is with only daemon part being different. Thanks again for being a resposive upstream.

**Pull request merged**